pymongo/pool.py

```diff
@@ -1375,7 +1375,8 @@ def connect(self, handler=None):
             if handler:
                 handler.contribute_socket(sock_info, completed_handshake=False)
 
-            sock_info.authenticate()
+            with self.lock:
+                sock_info.authenticate()
```

Comment:
Taking the pool lock here is a bit too coarse. The problem is that no other threads will be able to access the pool at all if any other thread is authenticating.

Instead we should introduce a new lock and only acquire it for the first step of MONGODB-AWS auth. Even that is a little more coarse than needed since it will serialize MONGODB-AWS auth with explicit creds too but it's a good start.

Reply:
I think it makes sense to put the lock in `auth_aws.py`, so we can use it for On-Demand KMS credentials as well, and it can utilize the `_create_lock` convention already present in this library.  Technically this will be too coarse as the lock will be held even for explicit credentials, but the lock will not be held for long in that case, since it is a simple environment lookup.

Reply:
We decided against locking, since it would interfere with the `maxConnecting` setting.